BUG: solve minimum_rotated_rectangle is incorrect in 2.0.0

idanmiara opened the pull request:
Closes https://github.com/shapely/shapely/issues/1670
restore and vectorized the custom `oriented_envelope` to solve [#1670](https://github.com/shapely/shapely/issues/1670)
add a new test and remove duplicated tests for `oriented_envelope` and `minimum_rotated_rectangle`

jorisvandenbossche reviewed (commented):
Thanks for updating the PR! Added a bunch of comments

idanmiara commented on shapely/algorithms/oriented_envelope.py:
```diff
@@ -0,0 +1,136 @@
+    geometry : Geometry or array_like
+    method: str or ..., optional
+        Determinates which method to use, supported methods:
+            * "min_area" - returns the oriented envelope with the minimal area (same as in Shapely<2), slower.
```
I gave it a name so users can choose this method regardless of possible change in the default ("best") method in future Shapely versions. 
WDYT ?

jorisvandenbossche replied:
It's already the default, and if we would change the implementation in the future (remove the python version in favor of a similar (fixed) one in GEOS), the "method" would still be the same (assuming the fixed GEOS version would also give the minimum area). So I currently don't see any reason why someone should do `method="min_area"` explicitly (and so why we should allow that)

idanmiara commented on shapely/tests/test_constructive.py:
```diff
@@ -901,74 +901,54 @@ def test_oriented_envelope_all_types(geometry):
-        ),
-    ],
+@pytest.mark.skipif(
+    shapely.minimum_rotated_rectangle == shapely.oriented_envelope,
```
Currently yes, maybe in some future version some decides they are not two names for the same function.
Otherwise do you want to test both functions or just one ?

jorisvandenbossche replied:
Yes, we want to test both (we want to ensure that both give the same result)

idanmiara commented on shapely/algorithms/oriented_envelope.py:
```diff
@@ -0,0 +1,136 @@
+    method: str or ..., optional
+        Determinates which method to use, supported methods:
+            * "min_area" - returns the oriented envelope with the minimal area (same as in Shapely<2), slower.
+            * ... - returns the oriented envelope using the GEOS implementation,
```
I agree that `method="geos"` is more clear than Ellipsis 👍

idanmiara commented on shapely/algorithms/oriented_envelope.py:
```diff
@@ -0,0 +1,136 @@
+    if geometry is None:
+        return None
+    if not hasattr(geometry, "geom_type"):
+        return np.array([oriented_envelope_min_area(g) for g in geometry])
```
I'm not sure what are suggesting to change here.
This inline should be faster than `np.vectorize` since it doesn't have the generalization overhead of `np.vectorized` which states `The [vectorize](https://numpy.org/doc/stable/reference/generated/numpy.vectorize.html#numpy.vectorize) function is provided primarily for convenience, not for performance. The implementation is essentially a for loop.`

jorisvandenbossche replied:
All the shapely ufuncs provide the full functionality of ufuncs, which among other things is that it works on arrays of any dimension, not just on 1D arrays. 
I made my comment above in the assumption that your code would only work for 1D arrays (seeing the single for loop), but missed that this is of course working recursively, and so will therefore also work on >1D arrays (the outer loop slices out one dimension, until you iterate over scalars).

So my comment wasn't for performance considerations, I know that in the end it's always a python loop. But numpy provides some helpers that turn a python function that works on scalar values into a fully compatible ufunc function. So it could still be useful to use that (but it might be `np.frompyfunc` and not `np.vectorize` that we want, not fully sure about the differences between both)

idanmiara replied:
> So my comment wasn't for performance considerations, I know that in the end it's always a python loop. But numpy provides some helpers that turn a python function that works on scalar values into a fully compatible ufunc function. So it could still be useful to use that (but it might be `np.frompyfunc` and not `np.vectorize` that we want, not fully sure about the differences between both)

On this specific case, that there is only single parameter to vectorized - the numpy wrappers wouldn't add any advantages. in general,  `np.vectorize` wraps `np.frompyfunc` to add even more generalized options (like auto deduction of input/output types), thus slower. So I don't think we should use either of them in this specific case, where we only iterate on a single argument.

idanmiara commented:
> Thanks for updating the PR! Added a bunch of comments

I addressed your comments (except https://github.com/shapely/shapely/pull/1708#discussion_r1103879221), then squashed and rebased as there were too many conflicts otherwise.

idanmiara commented:
Hmm... I'm not sure why it fails with something related to `LINESTRING Z`, there were not 3d geometries involved here...
```
Not equal to tolerance 1e-07
 x: array(<LINESTRING Z (1 1 0, 10 1 0)>, dtype=object)
 y: array(<LINESTRING (1 1, 10 1)>, dtype=object)
```

jorisvandenbossche reviewed (commented):
Thanks for the update! 
Added some additional comments. Further, there are some CI builds failing: the doctests are failing, and there seems to be an issue with Z dimension on older GEOS versions. But it's strange that it is actually the method="geos" version that is failing, and that should not have changed ...

jorisvandenbossche commented:
> > Hmm... I'm not sure why it fails with something related to `LINESTRING Z`, there were not 3d geometries involved here...
>  But it's strange that it is actually the method="geos" version that is failing, and that should not have changed ...

Ah, this is probably related to your change to use `assert_geometries_equal` instead of `equals(..).al()`. The latter is doing spatial equality, and will ignore the Z dimension. This seems to be a bug in older GEOS versions (incorrectly introducing a Z dimensio for this case) that we can ignore here.

idanmiara commented:
> Ah, this is probably related to your change to use `assert_geometries_equal` instead of `equals(..).al()`. The latter is doing spatial equality, and will ignore the Z dimension. This seems to be a bug in older GEOS versions (incorrectly introducing a Z dimensio for this case) that we can ignore here.

OK, I've updated the skip accordingly (3.6 -> 3.8)

caspervdw reviewed (approved):
Thanks for fixing this @jorisvandenbossche and @idanmiara . Implementation looks good, only personally I think it is unnecessary to test the “wrong” behaviour of the pre-GEOS 3.12 code. 

Maybe the function itself can be wrapped in a PRE_GEOS_312 conditional on the C level?

jorisvandenbossche commented:
> Maybe the function itself can be wrapped in a PRE_GEOS_312 conditional on the C level?

You mean so the ufunc is actually only available when built with GEOS 3.12? 
(This is a bit tricky, because we don't want to "officially" say that, but keeping it for all versions ensures that if someone really needs the ufunc but can't upgrade to the latest GEOS has it available through `shapely.lib.oriented_envelope`)

jorisvandenbossche commented on shapely/tests/test_constructive.py:
```diff
@@ -881,26 +881,23 @@ def test_oriented_envelope_all_types(geometry):
-            MultiPoint([(0, 0), (10, 0), (10, 10)]),
-            Polygon([[10.0, 10.0], [10.0, 0.0], [0.0, 0.0], [0.0, 10.0], [10.0, 10.0]]),
+            MultiPoint([(1.0, 1.0), (1.0, 5.0), (3.0, 6.0), (4.0, 2.0), (5.0, 5.0)]),
+            Polygon([(1.0, 1.0), (1.0, 6.0), (5.0, 6.0), (5.0, 1.0), (1.0, 1.0)]),
```
I changed the example here to be a more distinctive one. With this new example, there is a clear differentce in the "minimum area" and "minimum rectangle" solution. While with the previous input geometry, they were the same and there was actually two possible solutions with the same minimum area (the GEOS version vs our version then also returned a different one of those, I think only because of floating point differences or order of iteration)

jorisvandenbossche commented:
Going to move forward here, so we can make the release. 

But happy to further update the tests afterwards (or address other comments)

jorisvandenbossche commented:
Thanks @idan-miara for the initial PR!
